### ISSUE TITLE
Theory system overhaul

### ISSUE THREAD

**Issue by wwitzel** [PyProveIt/Prove-It#151]
There are a number of changes we've discussed that I'd like to implement soon.  Currently, this is my thinking (but is subject to change of course).

1. Change "context" to "theory".
2. Change "_theorems_" to "_theorems_and_conjectures_".
3. When a theorem/conjecture is displayed on the "_theorems_and_conjectures_" page, it should be annotated with a "status" of "conjecture without proof", "conjecture with conjecture-based proof", or "established theorem".
4. When importing axioms/theorems/conjectures (or common expressions), make it so we can simply import them at the theory package level so you don't have to specify _axioms_, or _theorems_and_conjectures_.  That means that the names have to be unique across the package -- that is a good thing to do in any case but we'll have to do some error checking on that.
5. When displaying a proof, use "conjecture" instead of "theorem" as appropriate (when it hasn't been fully proven).
6. Change "KnownTruth" to "ProvenTruth" -- it may be "proven" via "conjectures" so it isn't entirely "known".  Maybe there is a better name for this, but this is the best I've come up with so far.

**Comment by wdcraft01** [comment]
Some brief thoughts in response:

- Changing “context” to “theory”: I must admit to not having ever warmed to this idea, but I realize it might just be because I've grown accustomed to the “context” language. Does this mean we'll have a set theory theory? Or a group theory theory? Number theory theory? Or does it mean we'd simply then have “Set Theory,” “Group Theory,”  etc? Does it then mean in place of a logic context we'll have a “logic theory”? How about “real theory” in the number/sets/real context? It all sounds strange and awkward to me.
- Changing “theorems” to “theorems_and_conjectures”: seems reasonable.
- Annotations for when a theorem/conjecture is displayed … : seems reasonable and useful. What then would be the difference between “theorem” and “established theorem”? The “established” part is just re-emphasizing the non-conjecture status?
- Importing at context-level: could be convenient, but could also be annoying to not see/know more exactly where items are coming from and the fact that “from … import A, B” tells me nothing about the possibility that A represents a pre-defined literal (for example) while B represents a conjecture without proof.
- When displaying a proof … : might even be nice to indicate with some standardized notation the different conjecture statuses, perhaps something like “conjecture (w/p)” for a conjecture with conjecture-based proof, etc. We'd want it to be short and easily decipherable.
- Changing “KnownTruth” to “ProvenTruth”: I am very ambivalent about this and think it is worth some discussion. If a “ProvenTruth” could still be “proven” in the sense that it relies entirely on conjectures … not at all clear why we'd distinguish “known truth” from “proven truth.”

**Comment by wwitzel** [comment]
I edited my original post to add numbers to make it easier for this discussion.  Thanks for you input, Hoss.  Much appreciated.  Here are some responses.

1.  See Issue #54 .  My proposes is to use, for example, "theory of proveit.logic.sets".  I think that alleviates the awkwardness you mention, but do you still think it is problematic.
3.  Yeah, "established" is not strictly necessary.  I kind of like the clarity though.
4.  You make an excellent point that you want to know where to look when you import a common expression, axiom, or theorem/conjecture.  That is a problem.  The default would be to import from `_theorems_and_conjectures_` which is a bit lengthy.  We could drop the underscores when importing.  Hmm... not sure what to do on that one.
5.  Perhaps we could include the extra information, but having it both short and easily decipherable is a challenge.  I'm leaning toward not worrying about that.  If a user wants to check on the status of that particular "conjecture", they can click the link.
6.  I agree that "ProvenTruth" is not a slam dunk, but I do think it is an improvement.  It would be consistent with the terminology of 3: "conjecture with conjecture-based proof".  It's proven (has a proof) but the proof may be based on conjectures.

**Comment by wwitzel** [comment]
Regarding 3, although Hoss raised a valid objection, let me point out a couple of reasons why we still may want to do this.  First, his objection is mitigated somewhat by the fact that you can click on the expression of an axiom, theorem, or comment expression at it will take you to its definition.  Second, another reason to drop the `_axioms_` or `_theorems_and_conjections_` part is that I already do that in proofs (since it is redundant as the proof already identifies it as an axiom or theorem) and it would just make the imports be consistent with that.  This is something that has bugged me for a while and would be resolved if we implemented proposal 3.

**Comment by wwitzel** [comment]
Deepak suggested this a while ago, but I forgot the term:  "KnownTruth" should be "Sequent".  From an e-mail I just sent:

"Sequent" is the term that Deepak told is about before that seems to be the appropriate name for our "KnownTruth" objects.  According to the wikipedia page:
https://en.wikipedia.org/wiki/Sequent
A_1, ..., A_m |- B_1, ..., B_n
is a "sequent" that means that if A_1 and ... and A_m are true then B_1 or ... or B_n is true.

For us, we restrict n to be 1.  Also, to be clear on our interpretation, we would say
A_1, ..., A_m |- B
means that if A_1 is true and ... and A_m is true then B is true and none of these must be a boolean.

**Comment by wdcraft01** [comment]
In an email from @wwitzel from Sunday 28 June:

> I am having some doubts about "Sequent".  On this wikipedia page, it distinguishes natural deduction from sequent calculus: [https://en.wikipedia.org/wiki/Sequent_calculus](https://en.wikipedia.org/wiki/Sequent_calculus). It says that for “Natural deduction,” “Every (conditional) line has exactly one asserted proposition on the right.” And for "Sequent calculus", “Every (conditional) line has zero or more asserted propositions on the right.”
> If you then look at the "Natural deduction" section:
> https://en.wikipedia.org/wiki/Sequent_calculus#Natural_deduction_systems
> it refers to the "KnownTruth"-type thing as a "judgement".
> So, based on that, we should call it a "Judgement".  A "Sequent" is a generalization of a "Judgement" in which zero or more asserted propositions are on the right.  Should we call it "Judgement"?

My initial response was that I'm ok with “judgment,” and I see that this is keeping with some of Frege's original terminology and consistent with comments in [Martin-Löf's 1996 paper](https://ncatlab.org/nlab/files/MartinLofOnTheMeaning96.pdf), but I wish it didn't have the colloquial linguistic baggage of meaning something that is “subjective.”

Somewhat interestingly, I find “judgement” (with the “e” in the middle) looks odd/incorrect compared to “judgment” (without the middle “e”). But both are technically acceptable (though in American English, “judgment” seems to be the dominant form).

Some related semantic question(s): Does the judgment S &vdash; C indicate the proven form? Or does it make sense to say/write something like “we wish to prove that S &vdash; C”? For example, in some recently-added text in the current citation paper, we have the statement that “The goal … is to prove judgments to be true, or evident.” Is a judgment that hasn't been proven to be true still called a “judgment”? Would it possibly make more sense to use language such as “The goal … is to establish judgments of the form S &vdash; C” ?

Have we decided to embrace the “judgment” terminology?

**Comment by wwitzel** [comment]
Yes, I'd like to go with "judgment" or "judgement".  (The U.S./U.K. spelling discrepancies are annoying; I had a computer that kept defaulting to U.K. spell checking so I can't ever keep them straight).

In the code, a Judgment object will be created with a proof.  However, I think it is perfectly acceptable to refer to judgments that haven't been proven when we write or comment about it.  That's totally consistent with the way the term is used.  Prove-It only really cares about proven judgments but that shouldn't stop us from talking about unproven ones.  That does make the term more flexible than "KnownTruth" -- that's a good thing because it allows for metalogical discussions using the same terminology.

**Comment by wwitzel** [comment]
While we are overhauling the "theory system", let's change "containment" to "inclusion".

This came up in a recent conversation with Bob Carr where we decided that "contains" is appropriate language to be synonymous with "has as a member" and "includes" is appropriate language to be synonymous with "is a superset of".  We came across the following stack exchange post:
https://math.stackexchange.com/questions/729525/how-is-the-word-contains-defined-in-set-theory-in-relation-with-neighborhood
One of the posts says:

A useful convention, proposed by Paul Halmos (as far as I know), but unfortunately far from universally accepted, is to use "contains" for members and "includes" for subsets. Consider this answer propaganda for that convention.

This seems like a good convention and avoids ambiguity.  But to follow this, we need to change "containment" to "inclusion".
